src/Lean/Meta/Tactic/Grind/EMatchTheorem.lean

```diff
@@ -149,9 +149,11 @@ def isGenPattern? (pat : Expr) : Option (GenPatternInfo × Expr) :=
 
 /-- Returns `true` if `declName` is the name of a `match`-expression congruence equation. -/
 def isMatchCongrEqDeclName (declName : Name) : CoreM Bool := do
-  let declName := privateToUserName declName
   match declName with
-  | .str p s => return (← isMatcher p) && Match.isCongrEqnReservedNameSuffix s
+  | .str p s =>
+    pure (Match.isCongrEqnReservedNameSuffix s) <&&>
+      -- The matcher may or may not have the same visibility as the congruence equation.
+      (isMatcher p <||> isMatcher (privateToUserName p))
```

Comment:
@leodemoura This code broke when applied to private matcher declarations, does the change make sense?